core/src/main/java/discord4j/core/object/entity/channel/ForumChannel.java

```diff
@@ -0,0 +1,247 @@
+package discord4j.core.object.entity.channel;
+
+import discord4j.common.annotations.Experimental;
+import discord4j.core.GatewayDiscordClient;
+import discord4j.core.object.entity.ForumTag;
+import discord4j.core.object.reaction.DefaultReaction;
+import discord4j.core.spec.ForumChannelEditSpec;
+import discord4j.core.spec.StartThreadInForumChannelSpec;
+import discord4j.core.util.EntityUtil;
+import discord4j.discordjson.json.ChannelData;
+import reactor.core.publisher.Mono;
+
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * A Discord guild channel that contains organized threads with labels.
+ */
+@Experimental
+public final class ForumChannel extends BaseTopLevelGuildChannel implements CategorizableChannel {
+
+    public ForumChannel(GatewayDiscordClient gateway, ChannelData data) {
+        super(gateway, data);
+    }
+
+    /**
+     * Gets the topic for this guild forum.
+     *
+     * @return the user defined topic for this guild forum channel (0-4096 characters).
+     */
+    public Optional<String> getTopic() {
+        return getData().topic().toOptional()
+            .flatMap(opt -> opt); // Value is present on guild forum type but nullable
+    }
+
+    /**
+     * Gets whether this channel is considered NSFW (Not Safe For Work).
+     *
+     * @return {@code true} if this channel is considered NSFW (Not Safe For Work), {@code false} otherwise.
+     */
+    public boolean isNsfw() {
+        return getData().nsfw().toOptional().orElse(false);
+    }
+
+    /**
+     * Gets the channels {@link Flag} associated to this forum channel
+     * Unknown flags are currently ignored.
+     *
+     * @return An {@link EnumSet} representing the <b>known flags</b> for this forum channel.
+     */
+    public EnumSet<Flag> getFlags() {
+        return getData().flags().toOptional()
+            .map(Flag::valueOf)
+            .orElseThrow(IllegalStateException::new); // Mandatory for Forum Channels
+    }
+
+    /**
+     * Gets the default auto archive duration for threads in this forum channel
+     *
+     * @return Default client auto archive duration in minutes
+     */
+    public int getDefaultAutoArchiveDuration() {
+        return getData().defaultAutoArchiveDuration().toOptional()
+            .orElseThrow(IllegalStateException::new); // Mandatory for Forum channels
+    }
+
+    /**
+     * Gets the available forum tags for this forum channel
+     *
+     * @return A list containing every available forum tag for this channel
+     */
+    public List<ForumTag> getAvailableTags() {
+        return getData().availableTags().toOptional()
+            .orElseThrow(IllegalStateException::new) // Mandatory for Forum channels
```

Comment:
maybe i confuse but with this line if the Optional is empty ever throw the excepcion and not run the rest?

Reply:
Yes, this is to prevent inconsistent state.

Discord should always return a value, even an empty list. If we don't have the list, we have an illegal state.

Then, if we have no value (not an empty list), we throw an exception.

Reply:
then its not better make return a empty List rather than a exception?

Reply:
I consider returning an empty list is a risk of continuing to work with inconsistent state.

We can, but I don't think we should continue ?

Reply:
Maybe can be good anothers opinions here, i think its better a empty list, for automod i make this things for avoid fill of throws.

Reply:
I just tested if the `available_tags` field was present when there is no tag available. Currently, with the tests tags, we have this : 

```JSON
    "nsfw": false,
    "available_tags": [
        {
            "id": "1092466131622711328",
            "name": "Tag 1",
            "emoji_id": null,
            "emoji_name": "✅",
            "moderated": false
        },
        {
            "id": "1092466131622711329",
            "name": "Tag 2",
            "emoji_id": null,
            "emoji_name": "❤️",
            "moderated": false
        }
    ],
    "template": "",
    "default_reaction_emoji": null,
    "default_sort_order": null,
    "default_forum_layout": 1
```

Now, after deleting them : 
```JSON
    "nsfw": false,
    "available_tags": [],
    "template": "",
    "default_reaction_emoji": null,
    "default_sort_order": null,
    "default_forum_layout": 1
```

I think we can keep the throw mechanism if we have an inconsistent object. Otherwise, we should always have an empty array.